firebase-appdistribution/test-app/src/main/java/com/googletest/firebase/appdistribution/testapp/ScreenshotDetectionFeedbackTrigger.kt

```diff
@@ -0,0 +1,226 @@
+          return
+        }
+        seenImages.add(uri)
+        val path = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA))
+        info("Path: $path")
+        if (path.lowercase(Locale.getDefault()).contains("screenshot")) {
+          FirebaseAppDistribution.getInstance().startFeedback(infoTextResourceId, uri)
+        }
+      }
+    } catch (e: Exception) {
+      error("Could not determine if media change was due to taking a screenshot", e)
+    } finally {
+      cursor?.close()
+    }
+  }
+
+  private fun listenForScreenshots() {
+    currentActivity?.let { activity ->
+      activity.contentResolver.registerContentObserver(
+        MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
+        /* notifyForDescendants= */ true,
+        this
+      )
+    }
+  }
+
+  private fun stopListeningForScreenshots() {
+    currentActivity?.let { activity -> activity.contentResolver.unregisterContentObserver(this) }
+  }
+
+  // Other lifecycle methods
+  override fun onActivityDestroyed(activity: Activity) {}
+  override fun onActivityStarted(activity: Activity) {}
+  override fun onActivityStopped(activity: Activity) {}
+  override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
+
+  companion object {
+    const val TAG = "ScreenshotDetectionFeedbackTrigger"
+
+    private val permissionToRequest =
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) READ_MEDIA_IMAGES
+      else READ_EXTERNAL_STORAGE
+    private val shouldCheckIfPending = Build.VERSION.SDK_INT >= 29
+    private val contentProjection =
+      if (shouldCheckIfPending)
+        arrayOf(MediaStore.Images.Media.DATA, MediaStore.MediaColumns.IS_PENDING)
+      else arrayOf(MediaStore.Images.Media.DATA)
+
+    @SuppressLint("StaticFieldLeak") // Reference to Activity is set to null in onActivityPaused
+    private var instance: ScreenshotDetectionFeedbackTrigger? = null
+
+    /**
+     * Initialize the screenshot detection trigger for this application.
+     *
+     * This should be called during [Application.onCreate]. [enable] should then be called when you
+     * want to actually start detecting screenshots.
+     *
+     * @param application the [Application] object
+     * @param infoTextResourceId resource ID of info text to show to testers before giving feedback
+     */
+    open fun initialize(application: Application, infoTextResourceId: Int) {
+      if (instance == null) {
+        val handlerThread = HandlerThread("AppDistroFeedbackTrigger")
+        handlerThread.start()
+        instance = ScreenshotDetectionFeedbackTrigger(infoTextResourceId, Handler(handlerThread.looper))
+        application.registerActivityLifecycleCallbacks(instance)
+      }
+    }
+
+    /**
+     * Start listening for screenshots, and start feedback when a new screenshot is detected.
+     *
+     * @throws IllegalStateException if [initialize] has not been called yet
+     */
+    open fun enable() {
+      instance?.also {
```

Comment:
Curious: Why `also` instead of `let`?
Also (sic): what about just `if (instance == null) throw` as a "guard clause"?

Reply:
Similar to the smart cast issue I mentioned, we can't use smart cast here because this member may have been reassigned in between the null check and usage.

We could use `if (instance == null) throw` but we'd still have to use a `!!` to assert that it's non-null.

Here I use `?.also` instead of `?.let` because `?.also` is safer when you want to handle the `else` case. It evaluates to the value of the nullable input value, instead of `let` which evaluates to the return value of the block. For example, this code would actually throw an exception regardless of whether `myNullableVar` is null or not:

```
myNullableVar?.let {
  println("Hi")
  null
} ?: throw IllegalStateException("This will always be thrown")
```

I'm just going to go ahead and get rid of all uses of both, though. I don't think a couple saved lines is worth the ding to readability.

Reply:
Thanks for explaining!